### PR TITLE
Added native int64 indices support to gather layer

### PR DESCRIPTION
**Pull request by alexlyulkov**
Added int64 support for Gather layer indices without int64 to int32 conversion.

Also added minor fixes in Gather layer:
- removed useless indices mat reshape since indices mat is accessed as ptr
- removed preprocessing step (indices[i] = normalize_axis(indices[i]) and added normalize_axis to the line where an index is actually used

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [ ] There is a reference to the original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

**Review by fengyuentau — COMMENTED** [review]
New tests for int64 indices?

**Comment by alexlyulkov** [comment]
> New tests for int64 indices?

All existing tests use int32 or int64 indices, float indices are not allowed in dnn 5.x anymore.
Also there is a new Gather test with int indices and int values https://github.com/opencv/opencv/blob/5.x/modules/dnn/test/test_int.cpp#L273

File: modules/dnn/src/layers/gather_layer.cpp [hunk 1]
```diff
@@ -115,7 +102,7 @@ class GatherLayerImpl CV_FINAL : public GatherLayer
             const size_t src_offset = i * outer_size;
             for (size_t j = 0 ; j < indices_total; ++j)
             {
-                const int index = idx[j];
+                const int index = normalize_axis(idx[j], axis_size);
```
Comment (dkurt) [inline modules/dnn/src/layers/gather_layer.cpp:105]: As any type is casted to int anyway, I recommend to do before forward_impl. So it let's avoid using a template method.

Comment (alexlyulkov) [reply modules/dnn/src/layers/gather_layer.cpp:105]: As I understand, OpenCV supports only int32 Mat sizes, so there is no point to keep int64 index value.
ConvertTo creates a copy of Mat, so it uses some memory and CPU resources. Casting of a local variable doesn't uses additional memory and most likely doesn't affect performance.

**Review by vpisarev — APPROVED** [review]

**Pull request merged**